src/main/scala/com/fulcrumgenomics/util/ReadStructure.scala

```diff
@@ -0,0 +1,167 @@
+    this.filter(s => strict || s.offset < bases.length).map {
+      segment =>
+        val end = segment.offset + segment.length
+        if (bases.length < end) {
+          if (strict) throw new IllegalStateException("Not enough bases for read structure: " + this)
+          else (bases.substring(segment.offset), ReadSegment(segment, bases.length-segment.offset))
+        }
+        else (bases.substring(segment.offset, end), segment)
+    }
+  }
+
+  /** Splits the given bases and qualities into triples with its associated read segment. */
+  def extractWithQualities(bases: String, qualities: String, strict: Boolean = true): Seq[(String, String, ReadSegment)] = {
+    assert(bases.length == qualities.length)
+    this.filter(s => strict || s.offset < bases.length).map {
+      segment =>
+        val end = segment.offset + segment.length
+        if (strict) {
+          if (bases.length < end) throw new IllegalStateException("Not enough bases for read structure: " + this)
+          if (qualities.length < end) throw new IllegalStateException("Not enough qualities for read structure: " + this)
+        }
+        if (bases.length < end) (bases.substring(segment.offset), qualities.substring(segment.offset), ReadSegment(segment, bases.length-segment.offset))
+        else (bases.substring(segment.offset, end), qualities.substring(segment.offset, end), segment)
+    }
+  }
+
+  private def collectSegments[T <: ReadSegment: ClassTag]: Seq[T] = this.collect { case t: T => t }
+  def template: Seq[Template] = this.collectSegments[Template]
+  def sampleBarcode: Seq[SampleBarcode] = this.collectSegments[SampleBarcode]
+  def molecularBarcode: Seq[MolecularBarcode] = this.collectSegments[MolecularBarcode]
+  def skip: Seq[Skip] = this.collectSegments[Skip]
+}
+
+object ReadSegment {
+  val Types = Seq('T', 'B', 'M', 'S')
+
+  def apply(offset: Int, length: Int, s: String): ReadSegment = {
+    if (s.isEmpty) throw new IllegalArgumentException("Empty string given")
+    apply(offset, length, s.head)
+  }
+
+  def apply(offset: Int, length: Int, c: Char): ReadSegment = {
+    c match {
+      case 'T' => Template(offset, length)
+      case 'B' => SampleBarcode(offset, length)
+      case 'M' => MolecularBarcode(offset, length)
+      case 'S' => Skip(offset, length)
+      case _ => throw new IllegalArgumentException(s"Unknown read segment type: $c")
+    }
+  }
+
+
+  /** Creates a new read segment of the same type but with a new length */
+  def apply(segment: ReadSegment, length: Int): ReadSegment = segment match {
+    case seg: Template         => Template(seg.offset, length)
+    case seg: SampleBarcode    => SampleBarcode(seg.offset, length)
+    case seg: MolecularBarcode => MolecularBarcode(seg.offset, length)
+    case seg: Skip             => Skip(seg.offset, length)
+  }
+}
+
+/** Base for all types of segments of a read */
+sealed abstract class ReadSegment(val offset: Int, val length: Int, val symbol: Char) {
```

Comment:
Having this as an abstract class, and then having the subclasses be case classes that override the attributes feels strange - particularly if you override an attribute it's odd to then also have to pass the value to the super-class constructor.

How about:

```scala
sealed trait ReadSegment {
    def offset: Int
    def length: Int
    def symbol: Char
}
```

Then the `val`s in the case classes will fulfill the `def`s without the possible duplications?